### PR TITLE
Only recurse into incomplete folders during background scans

### PR DESCRIPTION
**Pull request by icewind1991**
Improves background scan performance on large folder structures

cc @PVince81 @rullzer

### CODE REVIEW

File: lib/public/files/cache/iscanner.php [hunk 1]
```diff
@@ -27,6 +27,7 @@
  * @since 9.0.0
  */
 interface IScanner {
+	const SCAN_RECURSIVE_INCOMPLETE = 2;
```
Comment (PVince81) [inline lib/public/files/cache/iscanner.php]: mind adding a PHP comment to explain the differences for future reference ?

Comment (PVince81) [inline lib/public/files/cache/iscanner.php]: the comment should probably go in the `scan()` method's PHPDoc

**Comment by PVince81** [comment]
Makes sense :+1:

File: lib/public/files/cache/iscanner.php (hunk 1 above)
Comment (icewind1991) [inline lib/public/files/cache/iscanner.php]: done

**Comment by icewind1991** [comment]
https://github.com/owncloud/core/pull/24702 makes this easier to test with

```
occ files:backgroundscan $user -v
```

you can see what folders the scanner goes into

**Comment by PVince81** [comment]
Second review ? @Xenopathic @mmattel

**Comment by PVince81** [comment]
@owncloud/filesystem @nickvergessen @schiesbn

File: lib/private/Files/Cache/Scanner.php [hunk 2]
```diff
@@ -386,6 +386,9 @@ protected function scanChildren($path, $recursive = self::SCAN_RECURSIVE, $reuse
 				if ($data) {
 					if ($data['mimetype'] === 'httpd/unix-directory' and $recursive === self::SCAN_RECURSIVE) {
 						$childQueue[$child] = $data;
+					} else if ($data['mimetype'] === 'httpd/unix-directory' and $recursive === self::SCAN_RECURSIVE_INCOMPLETE and $data['size'] === -1) {
```
Comment (rullzer) [inline lib/private/Files/Cache/Scanner.php]: why `and` and not `&&`?

**Comment by rullzer** [comment]
:+1:

**Comment by rullzer** [comment]
Conflicts.... Please rebase....

**Comment by icewind1991** [comment]
rebased

**Comment by PVince81** [comment]
Known failure due to issue on master -> merge

**Pull request merged**